### PR TITLE
Stable diffusion python API level perf improvements

### PR DESCRIPTION
**Pull request by AleksKnezevic**
Stable diffusion changes to get

- all group norms on device
- better performing cross attention
- (almost) no layout conversions between ops
- Most matmuls in bfp8

My apologies to all reviewers for the giant PR.

### CODE REVIEW

File: ttnn/ttnn/operations/core.py [hunk 1]
```diff
@@ -617,7 +617,13 @@ def _to_layout_validate_input_tensors(operation_name, input_tensor, *args, **kwa
 
 
 @ttnn.register_operation(name="ttnn.to_layout", validate_input_tensors=_to_layout_validate_input_tensors)
-def to_layout(tensor, layout: ttnn.Layout, dtype: ttnn.DataType = None):
+def to_layout(
+    tensor,
+    layout: ttnn.Layout,
+    dtype: ttnn.DataType = None,
+    output_memory_config: ttnn.MemoryConfig = None,
+    use_multicore: bool = False,
```
Comment (cfjchu) [inline ttnn/ttnn/operations/core.py:625]: why is this extra function parameter needed?

File: ttnn/ttnn/operations/normalization.py [hunk 2]
```diff
@@ -197,12 +199,81 @@ def rms_norm(input_tensor: ttnn.Tensor, weight: ttnn.Tensor, *, epsilon: float =
...
+    input_tensor = torch.nn.functional.pad(input_tensor, (0, zeros_to_insert))
+    input_tensor = input_tensor.flatten()
+    input_tensor = input_tensor[: num_channels + zeros_to_insert * (num_channels // values_per_chunk)]
+    return input_tensor.reshape(1, 1, -1, 32)
+
+
 def _torch_group_norm(input_tensor: ttnn.Tensor, *, num_groups, epsilon=1e-05, weight=None, bias=None, **_):
     import torch
 
+    print("start copy from dev to host")
```
Comment (cfjchu) [inline ttnn/ttnn/operations/normalization.py:270]: remove?

File: ttnn/ttnn/operations/normalization.py [hunk 3]
```diff
@@ -197,12 +199,81 @@ def rms_norm(input_tensor: ttnn.Tensor, weight: ttnn.Tensor, *, epsilon: float =
     return output_tensor
 
 
+# group norm helper function
+def determine_expected_group_norm_sharded_config_and_grid_size(
```
Comment (cfjchu) [inline ttnn/ttnn/operations/normalization.py:203]: @arakhmati to review this when he's back on Monday

File: ttnn/ttnn/types.py [hunk 4]
```diff
@@ -76,3 +76,14 @@ class ShardStrategy(Enum):
 class ShardOrientation(Enum):
     ROW_MAJOR = 1
     COLUMN_MAJOR = 2
+
+
+class ShardStrategy(Enum):
```
Comment (cfjchu) [inline ttnn/ttnn/types.py:81]: Please remove this. We already have these two classes defined in `ttnn`. See `ttnn/ttnn/__init__.py`. e.g. `ttnn.ShardStrategy.HEIGHT` should already work.

File: tt_eager/tensor/tensor.cpp [hunk 5]
```diff
@@ -411,7 +411,7 @@ Tensor create_sharded_device_tensor(const Shape& shape, DataType data_type, Layo
         TT_ASSERT((shard_shape[0] % TILE_HEIGHT == 0 && shard_shape[1] % TILE_WIDTH == 0), "Shard shape must be tile sized");
     } else if (layout == Layout::ROW_MAJOR) {
         // Require alignment for now
-        TT_ASSERT(shard_shape[1] * tensor_impl::element_size_bytes_wrapper(data_type) % ADDRESS_ALIGNMENT == 0);
+        TT_ASSERT(shard_shape[1] * tensor_impl::element_size_bytes_wrapper(data_type) % 16 == 0);
```
Comment (cfjchu) [inline tt_eager/tensor/tensor.cpp:414]: Original code here was `ADDRESS_ALIGNMENT` which is `32`.. Is the real alignment 16?

File: ttnn/ttnn/operations/core.py (hunk 1 above)
Comment (AleksKnezevic) [reply ttnn/ttnn/operations/core.py:625]: to allow you to set an output memory config and whether or not to use the multicore variant of the op.

File: ttnn/ttnn/operations/normalization.py (hunk 2 above)
Comment (AleksKnezevic) [reply ttnn/ttnn/operations/normalization.py:270]: good catch.

File: ttnn/ttnn/operations/transformer.py [hunk 6]
```diff
@@ -390,7 +390,6 @@ def attention_softmax(
             scaler,
             attention_mask,
             output_mem_config=memory_config,
-            program_config=program_config,
```
[the remark this answers is not in the document]
Comment (AleksKnezevic) [reply ttnn/ttnn/operations/transformer.py:393]: yes, this was a bug, that op does not take in a program config.

File: tt_eager/tensor/tensor.cpp (hunk 5 above)
Comment (AleksKnezevic) [reply tt_eager/tensor/tensor.cpp:414]: @tt-nshanker, you made this change. Thoughts?

Comment (tt-nshanker) [reply tt_eager/tensor/tensor.cpp:414]: For RM layout, the shard width should be 16 byte alignment not 32 which is required for TILE LAYOUT. @tarafdarTT please review this change. Thanks.

File: ttnn/ttnn/operations/normalization.py (hunk 3 above)
Comment (tt-nshanker) [reply ttnn/ttnn/operations/normalization.py:203]: Group norm should be a class in TTNN and this function should go there along with other helper functions for group norm. Is it okay to check this in for now? Once the GN class is added, this will be moved.

File: ttnn/ttnn/operations/core.py (hunk 1 above)
Comment (arakhmati) [reply ttnn/ttnn/operations/core.py:625]: Please rename `output_memory_config` to `memory_config`. We don't add the prefix `output` in `ttnn` because it's obvious it's for the output and makes the name shorter.

Comment (arakhmati) [reply ttnn/ttnn/operations/core.py:625]: Do we need `use_multicore` in `ttnn`? Can we just always set it to true when calling underlying `tt_lib` operations?

File: ttnn/ttnn/operations/normalization.py (hunk 3 above)
Comment (arakhmati) [reply ttnn/ttnn/operations/normalization.py:203]: Why does group norm need to be a class?

**Review by arakhmati — CHANGES_REQUESTED** [review]

File: ttnn/ttnn/operations/core.py (hunk 1 above)
Comment (tt-nshanker) [reply ttnn/ttnn/operations/core.py:625]: There is an issue with mutli core untilize. @tt-aho can confirm if we can turn on multi core by default in ttnn.to_layout function.

File: ttnn/ttnn/operations/normalization.py (hunk 3 above)
Comment (tt-nshanker) [reply ttnn/ttnn/operations/normalization.py:203]: GN takes in "config" tensors (masks for zeroing out portions of tiles) that are initialized at module building time and passed to GN op during runtime. This is similar to conv and maxpool that have a class to handle config tensors.

**Review by arakhmati — APPROVED** [review]
